File: AN_2/SEMESTRUL_IV/MPP/LABORATOR/JAVA_TEMA/tema_bun/Networking/src/main/java/com/example/tema_bun/objectProtocol/ServicesRpcProxy.java
package com.example.tema_bun.objectProtocol;

import com.example.tema_bun.IObserver;
import com.example.tema_bun.IServices;
import com.example.tema_bun.Utilizator;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.sql.SQLOutput;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingDeque;

import com.fasterxml.jackson.databind.ObjectMapper;

public class ServicesRpcProxy implements IServices {
    private String host;
    private int port;
    private ObjectInputStream input;
    private ObjectOutputStream output;
    private Socket connection;
    private BlockingQueue<Response> qresponses;
    private volatile boolean finished;
    private String username = null;
    private IObserver client = null;

    public ServicesRpcProxy(String host, int port) {
        this.host = host;
        this.port = port;
        qresponses = new LinkedBlockingDeque<Response>();
    }


    public boolean login(String username, String parola, IObserver client) throws Exception {
        initializeConnection();
        Utilizator utilizator=new Utilizator(username,parola);
        Request request = new Request.Builder().type(RequestType.LOGIN).data(utilizator).build();
        sendRequest(request);
        Response response = readResponse();

        Boolean loggedIn = false;

        if (response.type() == ResponseType.OK) {
            loggedIn = (Boolean) response.data();
            if (loggedIn) {
                this.client = client;
                return true;
            }

        } else if (response.type() == ResponseType.ERROR) {
            String err = response.data().toString();
            closeConnection();
            throw new Exception(err);
        }
        return loggedIn;
    }

    public void logout(String notused) throws Exception {
//        initializeConnection();
        Utilizator utilizator=new Utilizator(notused,null);
        Request req = new Request.Builder().type(RequestType.LOGOUT).data(utilizator).build();
        sendRequest(req);
        Response response = readResponse();
        closeConnection();
        username = null;
        client = null;
        if (response.type() == ResponseType.ERROR) {
            String err = response.data().toString();
            throw new Exception(err);
        }
    }

    public Utilizator getByAccount(String username, String parola){
        return null;
    }

    @Override
    public void createAccount(String usernamee, String parola) throws Exception{
        initializeConnection();
        Utilizator u=new Utilizator(usernamee,parola);
        Request req = new Request.Builder().type(RequestType.CREARE_CONT).data(u).build();
        sendRequest(req);
        Response response = readResponse();
        if (response.type() == ResponseType.ERROR) {
            String err = response.data().toString();
            throw new Exception(err);
        }
    }

    @Override
    public List<Integer> creareVectorCnt() throws Exception{
        initializeConnection();

        Request req = new Request.Builder().type(RequestType.GET_NUMBER_OF_PARTICIPANTS).build();
        sendRequest(req);
        Response response = readResponse();
            if (response.type() == ResponseType.ERROR) {
                String err = response.data().toString();
                throw new Exception(err);
            }

            return (List<Integer>) response.data();

    }

    public List<String> JoinInscrieriCopii(String proba,String categorie) throws Exception{
        initializeConnection();
        Request req = new Request.Builder().type(RequestType.GET_PROBA_CATEGORIE_PARTICIPANTS).data(proba+" "+categorie).build();
        sendRequest(req);
        Response response = readResponse();
        if (response.type() == ResponseType.ERROR) {
            String err = response.data().toString();
            throw new Exception(err);
        }
        List<String> rez=(List<String>)response.data();
        return (List<String>)response.data();
    }

    public void adaugaCopilInscriere(String nume, Integer varsta, List<String> probe,String usernameCurent) throws Exception
    {
//        initializeConnection();
        ObjectMapper objectMapper = new ObjectMapper();
        String[] inscriereData = new String[]{nume, varsta.toString(), objectMapper.writeValueAsString(probe),usernameCurent};

//        System.out.println(inscriereData[2]);

//        String data = objectMapper.writeValueAsString(inscriereData);

        Request req = new Request.Builder().type(RequestType.INSCRIERE).data(inscriereData).build();
        sendRequest(req);
        Response response = readResponse();
        if (response.type() == ResponseType.ERROR) {
            String err = response.data().toString();
            throw new Exception(err);
        }
    }


    private void sendRequest(Request request) throws Exception {
        try {
            output.writeObject(request);
            output.flush();
        } catch (Exception e) {
            System.out.println("intra pe eroare");
            throw new Exception("Error sending object "+e);
        }
    }

    private Response readResponse() throws Exception {
        Response response=null;
        try{
            response=qresponses.take();

        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        return response;
    }

    private void initializeConnection() {
        try {
            connection = new Socket(host, port);
            output=new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input=new ObjectInputStream(connection.getInputStream());
            finished=false;
            startReader();
        }catch (IOException e){
            e.printStackTrace();
        }
    }

    private void closeConnection() {
        finished=true;
        try {
            input.close();
            output.close();
            connection.close();
            client=null;
        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    private void startReader(){
        Thread tw=new Thread(new ReaderThread());
        tw.start();
    }



    private class ReaderThread implements Runnable {
        public void run() {
            while (!finished) {
                try {
                    Object response = input.readObject();
                    System.out.println("response received " + response);

                    if (((Response)response).type() == ResponseType.UPDATE)
                        client.participantInscris();
                    else {
                        try {
                            qresponses.put((Response) response);
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                    }
                } catch (IOException | ClassNotFoundException e) {
                    System.out.println("Reading error " + e);
                }
            }
        }
    }

}

File: AN_2/SEMESTRUL_IV/MPP/LABORATOR/CSHARP_TEMA/tema_sapt3/Networking/rpcprotocol/ServicesRpcProxy.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Threading;
using Model;
using Newtonsoft.Json;
using services;

namespace networking.rpcprotocol;

public class ServicesRpcProxy:IServices
{
    private string host;
    private int port;
    private Queue<Response> responses;
    private EventWaitHandle _waitHandle;
    
    private IFormatter formatter;
    private TcpClient connection;
    private NetworkStream stream;
    
    private volatile bool finished;
    private IObserver client;

    public ServicesRpcProxy(string host, int port)
    {
        this.host = host;
        this.port = port;
        responses=new Queue<Response>();
        /*initializeConnection();*/
    }

    
    
    
    
    private void sendRequest(Request request)
    {
        try
        {
            formatter.Serialize(stream, request);
            stream.Flush();
        }
        catch (Exception e)
        {
            throw new Exception("Error sending object "+e);
        }
    }
    
    
    private Response readResponse()
    {
        Response response =null;
        try
        {
            _waitHandle.WaitOne();
            lock (responses)
            {
                //Monitor.Wait(responses); 
                response = responses.Dequeue();
            }
        }
        catch (Exception e)
        {
            Console.WriteLine(e.StackTrace);
        }
        return response;
    }
    
    
    
    private void initializeConnection()
    {
        try
        {
            connection=new TcpClient(host,port);
            stream=connection.GetStream();
            formatter = new BinaryFormatter();
            finished=false;
            _waitHandle = new AutoResetEvent(false);
            startReader();
        }
        catch (Exception e)
        {
            Console.WriteLine(e.StackTrace);
        }
    }
    
    
    
    private void closeConnection()
    {
        finished=true;
        try
        {
            stream.Close();
            connection.Close();
            _waitHandle.Close();
            client=null;
        }
        catch (Exception e)
        {
            Console.WriteLine(e.StackTrace);
        }
    }
    
    
    private void startReader()
    {
        Thread tw =new Thread(run);
        tw.Start();
    }
    
    
    public virtual void run()
    {
        while(!finished)
        {
            try
            {
                Response response = (Response)formatter.Deserialize(stream);
                Debug.WriteLine("response received "+response);
                if (response.Type==ResponseType.UPDATE)
                {
                    Debug.WriteLine("am primit raspuns de update");
                    /*Console.WriteLine("am primit response de UPDATE");*/
                    client.participantInscris();
                    
                    /*AICI NU MAI AJUNGE DE CEEEEE???
                     ZICE CA E NULL POINTER CLIENT*/
                    Debug.WriteLine("am trecut de facut update");
                }
                else
                {
                    lock (responses)
                    {
                        responses.Enqueue((Response)response);
                    }
                    _waitHandle.Set();
                }
            }
            catch (Exception e)
            {
                Debug.WriteLine("Reading error "+e);
            }
					
        }
    }
    
    
    public bool login(string data, string password, IObserver clientObserver)
    {
        initializeConnection();
        Utilizator u=new Utilizator(data,password);
        Request request = new Request.Builder().type(RequestType.LOGIN).data(u).build();
        sendRequest(request);
        Response response = readResponse();

        Boolean loggedIn = false;

        if (response.Type == ResponseType.OK) {
            loggedIn = (Boolean) response.Data;
            if (loggedIn) {
                Debug.WriteLine("a logat clientul");
                this.client = clientObserver;
                return true;
            }
        } else if (
            response.Type == ResponseType.ERROR)
        {
            Debug.WriteLine("nu a logat clientul");
            string err = response.Data.ToString();
            closeConnection();
            throw new Exception(err);
        }
        return loggedIn;
    }

    
    
    public void logout(string email)
    {
        Utilizator u=new Utilizator(email,null);
        Request req = new Request.Builder().type(RequestType.LOGOUT).data(u).build();
        sendRequest(req);
        Response response = readResponse();
        closeConnection();
        client = null;
        if (response.Type == ResponseType.ERROR)
        {
            String err = response.Data.ToString();
            throw new Exception(err);
        }
    }

    public Utilizator getByAccount(string username, string parola)
    {
        throw new NotImplementedException();
    }

    public void createAccount(string username, string parola)
    {
        throw new NotImplementedException();
    }

    public List<int> creareVectorCnt()
    {
        /*throw new NotImplementedException();*/
        /*initializeConnection();*/
        Request req = new Request.Builder().type(RequestType.GET_NUMBER_OF_PARTICIPANTS).build();
        sendRequest(req);
        Response response = readResponse();
        /*closeConnection();*/
        /*client = null;*/
        if (response.Type == ResponseType.ERROR)
        {
            String err = response.Data.ToString();
            throw new Exception(err);
        }
        return (List<int>) response.Data;
    }

    public List<int> gasesteParticipanti(string proba, string categorie)
    {
        throw new NotImplementedException();
    }

    public string getNumeVarstaById(int id)
    {
        throw new NotImplementedException();
    }

    public List<string> gasesteNumeVarstaCopii(List<int> listCoduri)
    {
        throw new NotImplementedException();
    }

    public void adaugaCopilInscriere(string nume, int varsta, List<string> probe,String usernameCurent)
    {
        /*initializeConnection();*/
        String[] inscriereData = new String[]{nume, varsta.ToString(), JsonConvert.SerializeObject(probe),usernameCurent};

        Request req = new Request.Builder().type(RequestType.ADD_PARTICIPANT).data(inscriereData).build();
        sendRequest(req);
        Response response = readResponse();
        Console.WriteLine(response);
        if (response.Type == ResponseType.ERROR) {
            String err = response.Data.ToString();
            Console.WriteLine(response);
            throw new Exception(err);
        }
    }

    public List<String> JoinInscrieriCopii(String proba, String categorie)
    {
        /*initializeConnection();*/
        Request req = new Request.Builder().type(RequestType.GET_PROBA_CATEGORIE_PARTICIPANTS).data(proba+" "+categorie).build();
        sendRequest(req);
        Response response = readResponse();
        if (response.Type == ResponseType.ERROR) {
            String err = response.Data.ToString();
            throw new Exception(err);
        }
        List<String> rez=(List<String>)response.Data;
        return (List<String>)response.Data;
    }

    public Proba addProba(Proba proba) {
        return new Proba();
    }

    public Proba deleteProba(Proba proba)
    {
        return new Proba();
    }

    public List<Proba> getAllProba() {
        return new List<Proba>();
    }

    public Proba findByIdProba(String proba)
    {
        return new Proba();
    }
}